geopandas/tools/overlay.py

```diff
@@ -176,3 +180,204 @@ def overlay(df1, df2, how, use_sindex=True):
+    return dfsym
+
+
+def overlay_union(df1, df2):
+    """
+    Overlay Union operation used in overlay function
+    """
+    dfinter = overlay_intersection(df1, df2)
+    dfsym = overlay_symmetric_diff(df1, df2)
+    dfunion = pd.concat([dfinter, dfsym], ignore_index=True)
+    # keep geometry column last
+    columns = list(dfunion.columns)
+    columns.remove('geometry')
+    columns = columns + ['geometry']
+    return dfunion.reindex(columns=columns)
+
+
+def overlay(df1, df2, how='intersection', make_valid=True, reproject=True,
+            use_sindex=None):
+    """Perform spatial overlay between two polygons.
+
+    Currently only supports data GeoDataFrames with polygons.
+    Implements several methods that are all effectively subsets of
+    the union.
+
+    Parameters
+    ----------
+    df1 : GeoDataFrame with MultiPolygon or Polygon geometry column
+    df2 : GeoDataFrame with MultiPolygon or Polygon geometry column
+    how : string
+        Method of spatial overlay: 'intersection', 'union',
+        'identity', 'symmetric_difference' or 'difference'.
+    reproject : boolean, default True
+        If GeoDataFrames do not have same projection, reproject
+        df2 to same projection of df1 before performing overlay
+
+    Returns
+    -------
+    df : GeoDataFrame
+        GeoDataFrame with new set of polygons and attributes
+        resulting from the overlay
+
+    """
+    if use_sindex is not None:
+        warnings.warn("'use_sindex' is deprecated. The overlay operation "
+                      "always requires a spatial index (rtree).",
+                      DeprecationWarning, stacklevel=2)
+
+    # Allowed operations
+    allowed_hows = [
+        'intersection',
+        'union',
+        'identity',
+        'symmetric_difference',
+        'difference',  # aka erase
+    ]
+    # Error Messages
+    if how not in allowed_hows:
+        raise ValueError("`how` was '{0}' but is expected to be "
+                         "in %s".format(how, allowed_hows))
+
+    if isinstance(df1, GeoSeries) or isinstance(df2, GeoSeries):
+        raise NotImplementedError("overlay currently only implemented for "
+                                  "GeoDataFrames")
+
+    if (not df1.geom_type.isin(['Polygon', 'MultiPolygon']).all()
+            or not df2.geom_type.isin(['Polygon', 'MultiPolygon']).all()):
+        raise TypeError("overlay only takes GeoDataFrames with (multi)polygon "
+                        " geometries.")
+
+    # Computations
+    df1 = df1.copy()
+    df2 = df2.copy()
+    df1[df1._geometry_column_name] = df1.geometry.buffer(0)
+    df2[df2._geometry_column_name] = df2.geometry.buffer(0)
+    if df1.crs != df2.crs and reproject:
+        warnings.warn("Data has different projections. Converted data to "
+                      "projection of first GeoPandas GeoDataFrame",
+                      UserWarning)
+        df2.to_crs(crs=df1.crs, inplace=True)
```

Comment:
Is this safe? If `crs` do not match, would it not be safer to raise an exception? Seems like line 363 is making a decision that maybe only the user should make?

Reply:
Yes, that's a good question. I was thinking to leave this out for now. Also because currently the comparison of `crs` objects is currently very naive: it's simply a dict equality, and you can quickly end up with a slightly different dict which actually describes the same CRS.

Reply:
+1